src/imitation/algorithms/adversarial.py

```diff
@@ -233,45 +217,46 @@ def train_disc_step(
             provided, then take `n_gen` expert samples from the expert
             dataset, where `n_gen` is the number of samples in `gen_samples`.
             Observations should not be normalized.
+
+        Returns:
+           dict: Statistics for discriminator (e.g. loss, accuracy).
         """
         with logger.accumulate_means("disc"):
-            fetches = {
-                "train_op_out": self._disc_train_op,
-                "train_stats": self._discrim.train_stats,
-            }
             # optionally write TB summaries for collected ops
-            step = self._sess.run(self._global_step)
+            step = self._global_step
             write_summaries = self._init_tensorboard and step % 20 == 0
-            if write_summaries:
-                fetches["events"] = self._summary_op
 
-            # do actual update
-            fd = self._build_disc_feed_dict(
+            # compute loss
+            batch = self._make_disc_train_batch(
                 gen_samples=gen_samples, expert_samples=expert_samples
             )
-            fetched = self._sess.run(fetches, feed_dict=fd)
-
-            if write_summaries:
-                self._summary_writer.add_summary(fetched["events"], fetched["step"])
-
-            logger.logkv("step", step)
-            for k, v in fetched["train_stats"].items():
-                logger.logkv(k, v)
-            logger.dumpkvs()
+            disc_logits = self.discrim.logits_gen_is_high(
+                batch["state"],
+                batch["action"],
+                batch["next_state"],
+                batch["done"],
+                batch["log_policy_act_prob"],
+            )
+            loss = self.discrim.disc_loss(disc_logits, batch["labels_gen_is_one"])
 
-    def eval_disc_loss(self, **kwargs) -> float:
-        """Evaluates the discriminator loss.
+            # do gradient step
+            self._disc_opt.zero_grad()
+            loss.backward()
+            self._disc_opt.step()
 
-        Args:
-          gen_samples (Optional[rollout.Transitions]): Same as in `train_disc_step`.
-          expert_samples (Optional[rollout.Transitions]): Same as in
-            `train_disc_step`.
+            # compute/write stats and TensorBoard data
+            with th.no_grad():
+                train_stats = rew_common.compute_train_stats(
+                    disc_logits, batch["labels_gen_is_one"], loss
+                )
+            logger.record("step", step)
+            for k, v in train_stats.items():
+                logger.record(k, v)
+            logger.dump(self._global_step)
```

Comment:
Good catch on not incrementing `_global_step`—going to fix that.

> Confused about how both accumulate loggers use the same self._global_step. Aren't their steps meant to be different?

Is that a problem? I believe the step is only used for TensorBoard & human output, in which case it makes sense to use the same step for everything (at least under the assumption that everything is being written to the same TB file). Otherwise different graphs could be misaligned (e.g. step 10 in one graph happens before step 1 in another graph).

Reply:
I think we have it set up for different loggers logging to different event files: one for generator training, one for discriminator, one for an aggregation of both. The aggregated step is meant to be for each adversarial training epoch, but that could consist of multiple generator/discriminator updates.

Reply:
Adam's correct here, I'm going to update this before merge